File: python/counting-sort.py
from randomNumericList import randomNumericList

'''
Generate a count list with the initial values of zero.
'''
def getCountList(nums):
    maxi = max(nums)
    listCount = [0] * (maxi+1) #initialize the count array with all values as zero.
    
    for i in nums:
        listCount[i] = listCount[i] + 1
    
    return listCount

def countingSort(nums):
    listResult = []
    listCount = getCountList(nums)
    print("Count List: ", listCount)
    
    for i, value in enumerate(listCount):
        if value > 0:
            # Add the index values in the countList to the result list as many times as their count.
            for j in range(value):
                listResult.append(i)
    return listResult
    

nums = randomNumericList()
result =  countingSort(nums)
print("Result List: ", result)

File: php/counting-sort.php
<?php
include './randomNumericArray.php';

/**
 * Generate a count array with the initial values of zero.
 * @param array $nums: generated random numeric array
 * 
 * @return array count array
 */
function getCountArray($nums)
{
    $max = max($nums);
    $arrayCount = array_fill(0, $max + 1, 0); // initialize the count array with all values as zero.

    foreach ($nums as $value) {
        $arrayCount[$value] = $arrayCount[$value] + 1; //storing the count of each element at their indexes.
    }

    return $arrayCount;
}


function countingSort($nums)
{
    $arrayResult = [];
    $arrayCount = getCountArray($nums);
    echo "Count Array: ";
    print_r($arrayCount);


    foreach ($arrayCount as $i => $value) {
        if ($value > 0) {
            // Add the index values in the countArray to the result array as many times as their count.  
            for ($j = 0; $j < $value; $j++) {
                $arrayResult[] = $i;
            }
        }
    }
    return $arrayResult;
}

$nums = randomNumericArray(0, 10);
echo "Random Array: ";
print_r($nums);
$result = countingSort($nums);
echo "Result Array: ";
print_r($result);
